Fix replacing belongstomany junction records with complex types

markstory opened the pull request:
When the ORM was attempting to clear the junction records for a BelongsToMany junction table there were a few related problems causing it to fail:

1. The query used to locate the junction records did not have the   correct types set for where conditions.
2. If the junction table has a surrogate primary key that has    a different datatype than the association target table, the ORM    would use the target table type. In the committed fixtures this means    that the `id` column on the junction table would be typed as    `binaryuuid` instead of `integer`.

These factors combined together would result in either junction rows not being deleted. Thankfully the query to find junction records would always fail preventing more serious data corruption issues.

These problems can be solved by applying conditions *after* adding the junction join information and type mappings, and also setting junction table types for our aliased entity query.

I optimized object traversal inside the `_appendJunctionJoin` method as there were some redundant calls being ade.

Fixes #15212

markstory committed:
Fix replacing belongstomany junction records with complex types

When the ORM was attempting to clear the junction records for
a BelongsToMany junction table there were a few related problems causing
it to fail:

1. The query used to locate the junction records did not have the
   correct types set for where conditions.
2. If the junction table has a surrogate primary key that has
   a different datatype than the association target table, the ORM
   would use the target table type. In the committed fixtures this means
   that the `id` column on the junction table would be typed as
   `binaryuuid` instead of `integer`.

These factors combined together would result in either junction rows not
being deleted. Thankfully the query to find junction records would
always fail preventing more serious data corruption issues.

These problems can be solved by applying conditions *after* adding the
junction join information and type mappings, and also setting junction
table types for our aliased entity query.

I optimized object traversal inside the `_appendJunctionJoin` method
as there were some redundant calls being made.

Fixes #15212

othercorey commented on tests/Fixture/BinaryUuiditemsBinaryUuidtagsFixture.php:
```diff
@@ -0,0 +1,48 @@
+            'primary' => ['type' => 'primary', 'columns' => ['id']],
+            'unique_item_tag' => [
+                'type' => 'unique',
+                'columns' => ['binary_uuiditem_id', 'binary_uuidtag_id']
+            ]
```
```suggestion
                'columns' => ['binary_uuiditem_id', 'binary_uuidtag_id'],
            ],
```

othercorey committed:
Update tests/Fixture/BinaryUuiditemsBinaryUuidtagsFixture.php

ADmad commented:
The failure for SQL Server seems related to these changes.

markstory committed:
Try a different style of find() to see if SQLServer will accept it.

markstory committed:
Try a different approach to appease SQL server.

markstory committed:
Skip test on sqlserver.

Skip the binary uuid belongstomany test on sqlserver. It is failing in
hard to diagnose ways and I'd like to get a fix out for the other more
frequently used databases as there is a release planned for later today.
I can revisit this test when I figure out how to get sqlserver working
in linux.

othercorey commented on tests/TestCase/ORM/Association/BelongsToManyTest.php:
```diff
@@ -995,6 +1003,47 @@ public function testReplaceLinkFailingDomainRules()
+    public function testReplaceLinkBinaryUuid()
+    {
+        $this->skipIf(
+            $this->connection->getDriver() instanceof \Cake\Database\Driver\Sqlserver,
```
```suggestion
            ConnectionManager::get('test')->getDriver() instanceof \Cake\Database\Driver\Sqlserver,
```

markstory committed:
Update tests/TestCase/ORM/Association/BelongsToManyTest.php

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>
